lib/expressions/Term.ts

```diff
@@ -91,38 +96,57 @@ export class Literal<T> extends Term {
-  private static readonly specificFormatters: {[key in C.PrimitiveNumericType]: (val: number) => string } = {
+  private readonly specificFormatter: ((val: number) => string);
+  public constructor(
+    public typedValue: number,
+    public typeURL: RDF.NamedNode,
+    public strValue?: string,
+    public language?: string,
+  ) {
+    super(typedValue, typeURL, strValue, language);
+    if (!typeCanBeProvidedTo(<LiteralTypes> typeURL.value, TypeAlias.SPARQL_NUMERIC)) {
+      throw new Error(
+        `TypeUrl '${this.type}' provided to NumericLiteral should implement ${TypeAlias.SPARQL_NUMERIC}`,
+      );
+    }
+
+    const type = this.type;
     // Avoid emitting non lexical integers
-    integer: value => value.toFixed(0),
-    float: value => value.toString(),
-    decimal: value => value.toString(),
-    // // Be consistent with float
-    // decimal: (value) => {
-    //   const jsDecimal = value.toString();
-    //   return jsDecimal.match(/\./)
-    //     ? jsDecimal
-    //     : jsDecimal + '.0';
-    // },
-
-    // https://www.w3.org/TR/xmlschema-2/#double
-    double(value) {
-      const jsExponential = value.toExponential();
-      const [ jsMantisse, jsExponent ] = jsExponential.split('e');
-
-      // Leading + must be removed for integer
-      // https://www.w3.org/TR/xmlschema-2/#integer
-      const exponent = jsExponent.replace(/\+/u, '');
-
-      // SPARQL test suite prefers trailing zero's
-      const mantisse = jsMantisse.includes('.') ?
-        jsMantisse :
-        `${jsMantisse}.0`;
-
-      return `${mantisse}E${exponent}`;
-    },
-  };
-
-  public type: C.PrimitiveNumericType;
+    if (typeCanBeProvidedTo(type, TypeURL.XSD_INTEGER)) {
+      this.specificFormatter = value => value.toFixed(0);
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DECIMAL)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_FLOAT)) {
+      this.specificFormatter = value => value.toString();
+    } else if (typeCanBeProvidedTo(type, TypeURL.XSD_DOUBLE)) {
+      // https://www.w3.org/TR/xmlschema-2/#double
+      this.specificFormatter = value => {
+        const jsExponential = value.toExponential();
+        const [ jsMantisse, jsExponent ] = jsExponential.split('e');
+
+        // Leading + must be removed for integer
+        // https://www.w3.org/TR/xmlschema-2/#integer
+        const exponent = jsExponent.replace(/\+/u, '');
+
+        // SPARQL test suite prefers trailing zero's
+        const mantisse = jsMantisse.includes('.') ?
+          jsMantisse :
+          `${jsMantisse}.0`;
+
+        return `${mantisse}E${exponent}`;
+      };
+      // // Be consistent with float
+      // decimal: (value) => {
+      //   const jsDecimal = value.toString();
+      //   return jsDecimal.match(/\./)
+      //     ? jsDecimal
+      //     : jsDecimal + '.0';
+      // },
+    }
+  }
+
+  // ExtensionTable[type][TypeAlias.SPARQL_NUMERIC] !== undefined
```

Comment:
Yes, that's not clear at all. I wanted to have some way of telling this type was a `SPARQL_NUMERIC` but I ended up checking this in the constructor so it can go.